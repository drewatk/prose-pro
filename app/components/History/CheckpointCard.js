import React from "react";
import { connect } from "react-redux";
import * as R from "ramda";
import {
  Card,
  CardBody,
  CardSubtitle,
  CardText,
  Dropdown,
  DropdownToggle,
  DropdownMenu,
  DropdownItem
} from "reactstrap";
import moment from "moment";

import styles from "./CheckpointCard.scss";

import { EditorState, convertFromRaw } from "draft-js";
import { stateToHTML } from "draft-js-export-html";

import diff from "app/utils/diff";
import {
  UPDATE_EDITOR_STATE,
  SET_VIEW_STATE,
  SET_EDIT_STATE,
  SET_DIFF_STATE
} from "app/actions/editor";
import updateHistory from "app/actions/history";

import { FileObject } from "app/git-abs/metadata/file-object";

import { showError } from "app/actions/error";

export class CheckpointCard extends React.Component {
  constructor(props) {
    super(props);

    this.toggle = this.toggle.bind(this);
    this.state = {
      dropdownOpen: false
    };
  }

  toggle() {
    this.setState(prevState => ({
      dropdownOpen: !prevState.dropdownOpen
    }));
  }

  render() {
    const {
      version,
      timestamp,
      commit,
      gitAbstractions,
      currentFile,
      checkpointHistory,
      viewedCheckpoint,
      dispatch
    } = this.props;

    return (
      <Card
        className={`${styles.card} ${
          viewedCheckpoint === commit ? styles.cardViewed : ""
        }`}
        data-test-id="history-list-card"
      >
        <CardBody>
          <CardSubtitle>Checkpoint:</CardSubtitle>
          <CardText>{version}</CardText>
          <CardSubtitle>Date:</CardSubtitle>
          <CardText>{moment(timestamp).fromNow()}</CardText>
          <Dropdown isOpen={this.state.dropdownOpen} toggle={this.toggle}>
            <DropdownToggle tag="span">
              <i
                className="fas fa-ellipsis-h"
                style={{ float: "right" }}
                data-test-id="history-list-card-toggle"
              />
            </DropdownToggle>
            <DropdownMenu>
              <DropdownItem
                onClick={() => {
                  gitAbstractions
                    .switchToCurrentVersion(currentFile)
                    .then(() =>
                      gitAbstractions.switchVersion(currentFile, commit)
                    )
                    .then(fileData =>
                      dispatch({
                        type: UPDATE_EDITOR_STATE,
                        payload: EditorState.createWithContent(
                          convertFromRaw(fileData)
                        )
                      })
                    )
                    .then(() =>
                      gitAbstractions.switchToCurrentVersion(currentFile)
                    )
                    .then(() =>
                      dispatch({ type: SET_VIEW_STATE, payload: commit })
                    )
                    .catch(e => {
                      console.error("Error in Checkpoint View: ", e);
                      dispatch(showError(e.message));
                    });
                }}
                data-test-id="history-list-card-view-button"
              >
                View
              </DropdownItem>
              <DropdownItem
                onClick={() => {
                  gitAbstractions
                    .reset(currentFile, commit)
                    .then(fileData =>
                      dispatch({
                        type: UPDATE_EDITOR_STATE,
                        payload: EditorState.createWithContent(
                          convertFromRaw(fileData)
                        )
                      })
                    )
                    .then(() => dispatch({ type: SET_EDIT_STATE }))
                    .then(() =>
                      gitAbstractions.updateVersions(
                        currentFile,
                        new FileObject({
                          versions: R.dropLastWhile(
                            cp => cp.commit !== commit,
                            checkpointHistory
                          )
                        })
                      )
                    )
                    .then(() => gitAbstractions.getVersions(currentFile))
                    .then(({ versions }) => dispatch(updateHistory(versions)))
                    .catch(err => {
                      console.error("Reset Failed: ", err);
                      dispatch(showError(err.message));
                    });
                }}
                data-test-id="history-list-card-revert-button"
              >
                Revert
              </DropdownItem>
              <DropdownItem
                onClick={() => {
                  let selectedVersion = null,
                    previousVersion = null;
                  gitAbstractions
                    .switchToCurrentVersion(currentFile)
                    .then(() =>
                      gitAbstractions.switchVersion(currentFile, commit)
                    )
                    .then(selectedVersionData => {
                      selectedVersion = stateToHTML(
                        convertFromRaw(selectedVersionData)
                      );
                      return gitAbstractions.getVersions(currentFile);
                    })
                    .then(({ versions }) => {
                      const previousCommit = R.last(
                        R.takeWhile(v => v.commit !== commit, versions)
                      );
                      if (previousCommit) {
                        return gitAbstractions
                          .switchToCurrentVersion(currentFile)
                          .then(() =>
                            gitAbstractions.switchVersion(
                              currentFile,
                              previousCommit.commit
                            )
                          );
                      }
                      // return an empty file
                      return null;
                    })
                    .then(previousVersionData => {
                      if (previousVersionData === null) {
                        previousVersion = "";
                      } else {
                        previousVersion = stateToHTML(
                          convertFromRaw(previousVersionData)
                        );
                      }

                      return gitAbstractions.switchToCurrentVersion(
                        currentFile
                      );
                    })
                    .then(() =>
                      dispatch({
                        type: SET_DIFF_STATE,
                        payload: diff(previousVersion, selectedVersion).join("")
                      })
                    )
                    .catch(e => console.error("Error in Checkpoint Diff: ", e));
                }}
                data-test-id="history-list-card-diff-button"
              >
                Diff
              </DropdownItem>
            </DropdownMenu>
          </Dropdown>
        </CardBody>
      </Card>
    );
  }
}

const mapStateToProps = ({
  gitAbstractions,
  currentFile,
  checkpointHistory,
  editor: { viewedCheckpoint }
}) => ({
  gitAbstractions,
  currentFile,
  checkpointHistory,
  viewedCheckpoint
});

export default connect(mapStateToProps)(CheckpointCard);
