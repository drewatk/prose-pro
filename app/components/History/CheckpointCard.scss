.card {
  border: none !important;
  background: var(--dark-white-color) !important;
  color: var(--accent-color) !important;
  font-size: 1.2rem !important;
  margin-bottom: 5px;
}

.card-viewed {
  background: var(--accent-color) !important;
  color: var(--white-color) !important;
}

:global(.card-body) {
  padding: 10px !important;
}

:global(.card-subtitle) {
  margin-top: 0 !important;
  font-weight: bold !important;
}

:global(.card-text) {
  margin-bottom: 5px !important;
  font-size: 0.9rem !important;
}

:global(.dropdown-menu) {
  background-color: var(--accent-color) !important;
  border-color: var(--dark-white-color) !important;
}

:global(.dropdown-item) {
  color: var(--white-color) !important;

  &:hover {
    background-color: var(--dark-white-color) !important;
    color: var(--dark-color) !important;
  }
}
