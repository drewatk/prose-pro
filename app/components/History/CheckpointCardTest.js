import React from "react";

import {
  Card,
  CardBody,
  CardSubtitle,
  CardText,
  Dropdown,
  DropdownToggle,
  DropdownMenu,
  DropdownItem
} from "reactstrap";

class CheckpointCard extends React.Component {
  constructor(props) {
    super(props);

    this.toggle = this.toggle.bind(this);
    this.state = {
      dropdownOpen: false
    };
  }

  toggle() {
    this.setState(prevState => ({
      dropdownOpen: !prevState.dropdownOpen
    }));
  }

  render() {
    const { version, timestamp } = this.props;

    return (
      <Card>
        <CardBody>
          <CardSubtitle>Checkpoint:</CardSubtitle>
          <CardText>{version}</CardText>
          <CardSubtitle>Date:</CardSubtitle>
          <CardText>
            {new Date(timestamp).toLocaleString("en-US", {
              timeZone: "America/New_York"
            })}
          </CardText>
          <Dropdown isOpen={this.state.dropdownOpen} toggle={this.toggle}>
            <DropdownToggle tag="span">
              <i className="fas fa-ellipsis-h" style={{ float: "right" }} />
            </DropdownToggle>
            <DropdownMenu>
              <DropdownItem>View</DropdownItem>
              <DropdownItem>Revert</DropdownItem>
            </DropdownMenu>
          </Dropdown>
        </CardBody>
      </Card>
    );
  }
}

export default CheckpointCard;
