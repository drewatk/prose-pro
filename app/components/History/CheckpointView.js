import React from "react";
import { Card, CardBody, CardSubtitle, CardText } from "reactstrap";

const CheckpointView = ({ message, date }) => (
  <Card>
    <CardBody>
      <CardSubtitle>Checkpoint:</CardSubtitle>
      <CardText>{message}</CardText>
      <CardSubtitle>Date:</CardSubtitle>
      <CardText>{date}</CardText>
    </CardBody>
  </Card>
);

export default CheckpointView;
