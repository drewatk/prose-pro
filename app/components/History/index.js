import React from "react";
import { connect } from "react-redux";
import * as R from "ramda";

import CheckpointCard from "./CheckpointCard";

export const History = ({ history }) => (
  // TODO: Move styles to CSS file
  <div style={{ height: "100%" }}>
    {R.reverse(history).map((data, index) => (
      <CheckpointCard key={index} {...data} />
    ))}
  </div>
);

const mapStateToProps = ({ checkpointHistory }) => ({
  history: checkpointHistory
});

export default connect(mapStateToProps)(History);
