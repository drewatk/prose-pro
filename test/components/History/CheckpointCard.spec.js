import React from "react";
import Enzyme, { shallow, mount } from "enzyme";
import Adapter from "enzyme-adapter-react-16";
import { CheckpointCard } from "app/components/History/CheckpointCard";
import moment from "moment";

jest.mock("moment");

moment.mockImplementation(timestamp => ({
  fromNow: jest.fn(() => `moment_mock_fromNow_${timestamp}`)
}));

Enzyme.configure({ adapter: new Adapter() });

describe("CheckpointCard", () => {
  it("matches snapshot", () => {
    const props = { version: "test message", timestamp: 10123 };
    const wrapper = shallow(<CheckpointCard {...props} />);
    expect(wrapper).toMatchSnapshot();
  });

  it("shows dropdown on click", () => {
    const props = { version: "test message", timestamp: 10123 };
    const wrapper = mount(<CheckpointCard {...props} />);

    wrapper.find("i").simulate("click");

    expect(wrapper.state("dropdownOpen")).toBeTruthy();
  });
});
