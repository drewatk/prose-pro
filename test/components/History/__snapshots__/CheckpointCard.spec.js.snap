// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CheckpointCard matches snapshot 1`] = `
<Card
  className="card cardViewed"
  data-test-id="history-list-card"
  tag="div"
>
  <CardBody
    tag="div"
  >
    <CardSubtitle
      tag="h6"
    >
      Checkpoint:
    </CardSubtitle>
    <CardText
      tag="p"
    >
      test message
    </CardText>
    <CardSubtitle
      tag="h6"
    >
      Date:
    </CardSubtitle>
    <CardText
      tag="p"
    >
      moment_mock_fromNow_10123
    </CardText>
    <Dropdown
      active={false}
      addonType={false}
      direction="down"
      inNavbar={false}
      isOpen={false}
      nav={false}
      setActiveFromChild={false}
      toggle={[Function]}
    >
      <DropdownToggle
        aria-haspopup={true}
        color="secondary"
        tag="span"
      >
        <i
          className="fas fa-ellipsis-h"
          data-test-id="history-list-card-toggle"
          style={
            Object {
              "float": "right",
            }
          }
        />
      </DropdownToggle>
      <DropdownMenu
        flip={true}
        tag="div"
      >
        <DropdownItem
          data-test-id="history-list-card-view-button"
          onClick={[Function]}
          tag="button"
          toggle={true}
        >
          View
        </DropdownItem>
        <DropdownItem
          data-test-id="history-list-card-revert-button"
          onClick={[Function]}
          tag="button"
          toggle={true}
        >
          Revert
        </DropdownItem>
        <DropdownItem
          data-test-id="history-list-card-revert-button"
          onClick={[Function]}
          tag="button"
          toggle={true}
        >
          Diff
        </DropdownItem>
      </DropdownMenu>
    </Dropdown>
  </CardBody>
</Card>
`;
